File: Témata/04 nelineární regresní model (M-H algoritmus)/Matlab/cv05_pr02b.m
clear
close all
clc

%% Nastaveni Random-Walk Chain M-H algoritmu
S = 50000+1; %celkovy pocet generovanych vzorku + pocatecni vzorek
S0 = 30000+1; %celkovy pocet vyhozenych vzorku
S1 = S-S0; %pocet ponechanych vzorku k dalsi analyze

theta = zeros(1,S); %radkovy vektor pro uchovavani vzorku
theta(1) = 0; %pocatecni vzorek - theta(0)
count = 0; %pocitadlo poctu akceptovanych vzorku

% kandidatska hustota N(theta(s-1),c^2)
c = 5;

%% Random Walk Chain M-H algoritmus
for s=2:S
   %a) generovani kandidata theta_ast
   theta_ast = randn*c+theta(s-1);
   
   %b) spocitame akceptacni pravdepodobnost
   alpha = min(exp(-1/2*abs(theta_ast) ...
               +1/2*abs(theta(s-1))),1);
   
   %c) rozhodnuti o akceptaci nebo zamitnuti kandidata
   if alpha > rand
       theta(s) = theta_ast;
       count = count+1;
   else
       theta(s) = theta(s-1);
   end
end

%% Prezentace vysledku
theta(1:S0) = []; %vyhozeni prvnich S0 vzorku

%graficke zobrazeni konvergence
figure
plot(theta(1:500:end))

%pocitani momentu
E_theta = mean(theta);
D_theta = mean(theta.^2)-E_theta^2;
avg_count = count/(S-1); %prumerna mira akceptace

%zobrazeni vysledku na obrazovku
fprintf('E_theta   D_theta  Prumerna akceptace \n')
fprintf('%6.4f    %6.4f   %6.4f\n', E_theta, D_theta,avg_count)

%Histogram rozdeleni
figure
histogram(theta,50)

File: 04_nelinearni_regresni_model_m-h_algoritmus/cv05_pr02b.py
# Nastaveni cesty do domovske slozky
import sys
root_dir = "/".join(sys.argv[0].split("/")[:-3])
sys.path.insert(1, root_dir + "/Support/Python")

# importy pomocnych funkci
from progress_info import progress_bar

import numpy as np
import math
from tabulate import tabulate
import matplotlib.pyplot as plt

# ===== 1. Nastaveni independence chain M-H algoritmu =====

s = 50_000 + 1
s_0 = 30_000 + 1

theta = [0] * s
count = 0

# Kandidatska hustota N(theta[i-1], c**2)
c = 5

# ===== 2. Random walk chain M-H algoritmus =====

print("Random walk chain M-H algoritmus:")
for i in range(1, s):
    # a) Generovani kandidata theta_ast
    theta_ast = np.random.normal() * c + theta[i - 1]
    # b) Spocitani akceptacni pravdepodobnosti
    alpha = min(math.exp( -1/2 * abs(theta_ast) + 1/2 * abs(theta[i-1])), 1)
    # c) rozhodnuti o prijeti nebo odmitnuti kandidata
    if alpha > np.random.rand():
        theta[i] = theta_ast
        count += 1
    else:
        theta[i] = theta[i - 1]

    progress_bar(i, s)

# ===== 3. Prezentace vysledku =====
theta = theta[s_0:]     # Vyhozeni prvnich s_0 vzorku

# Graficke zobrazeni konvergence
fig, axs = plt.subplots(1,2, tight_layout=True)

axs[0].plot(theta[::500])
axs[0].set_title("Kazdy 500. odhad theta")
axs[1].hist(theta, bins=50, facecolor="blue", alpha=0.5, edgecolor="black")
axs[1].set_title("Rozdeleni theta")

# Pocitani momentu
e_theta = np.mean(theta)
d_theta = np.var(theta)
avg_count = count / (s - 1)     # Prumerna mira akceptace

headers = ["E[theta]", "D[theta]", "Prumerna akceptace"]
table = [[round(e_theta, 4), round(d_theta, 4), round(avg_count, 4)]]
print(tabulate(table, headers, tablefmt="pretty"))

plt.show()
